numba/tests/test_function.py

```diff
@@ -0,0 +1,575 @@
+import types as pytypes
+from numba import njit, function, cfunc, types, int64, int32
+import ctypes
+
+from .support import TestCase
+
+
+def dump(foo):  # FOR DEBUGGING, TO BE REMOVED
+    foo_type = function.fromobject(foo)
+    foo_sig = foo_type.signature()
+    foo.compile(foo_sig)
+    print('{" LLVM IR OF "+foo.__name__+" ":*^70}')
+    print(foo.inspect_llvm(foo_sig.args))
+    print('{"":*^70}')
+
+
+# Decorators for tranforming a Python function to different kinds of functions:
+def pure_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    func.pyfunc = func
+    return func
+
+
+def cfunc_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    f = cfunc(sig)(func)
+    f.pyfunc = func
+    return f
+
+
+def njit_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    f = njit(func)
+    f.pyfunc = func
+    return f
+
+
+def njit2_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    f = njit(sig)(func)
+    f.pyfunc = func
+    return f
+
+
+def ctypes_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    cfunc = cfunc_func(func, sig)
+    addr = cfunc._wrapper_address
+    if sig == int64(int64):
+        f = ctypes.CFUNCTYPE(ctypes.c_int64)(addr)
+        f.pyfunc = func
+        return f
+    raise NotImplementedError(
+        f'ctypes decorator for {func} with signature {sig}')
+
+
+class WAP(types.WrapperAddressProtocol):
+
+    def __init__(self, func, sig):
+        self.pyfunc = func
+        self.cfunc = cfunc_func(func, sig)
+        self.sig = sig
+
+    def __wrapper_address__(self, sig):
+        assert self.sig == sig, (self.sig, sig)
+        return self.cfunc._wrapper_address
+
+    def signature(self):
+        return self.sig
+
+    def __call__(self, *args, **kwargs):
+        return self.pyfunc(*args, **kwargs)
```

Comment:
Should this be available as a mixin for public use?

Reply:
Perhaps. `WAP` provides a way to turn any Python function to a first-class function as long as decorating it with `cfunc` makes sense.

Reply:
Actually, this issue is about supporting pure Python functions as first-class function type objects and is a little out-of-scope of this PR that introduces the feature mostly in terms of `cfunc` compiled functions. I'd like to postpone discussing this feature to the next PR as I think there is a cleaner way to support pure Python functions.